### ISSUE TITLE
Error response from daemon: is vmnetd running?

### ISSUE THREAD

**Issue by rbouma** [docker/for-mac#6677]
- [x] I have tried with the latest version of Docker Desktop
  - [x] I have tried disabling enabled experimental features
  - [x] I have uploaded Diagnostics
  - Diagnostics ID: 1890E3D2-2574-45D2-87A9-131871CF2E43/20230113085338

### Expected behavior

When running  our local devstack it needs to be binding to port 53 (priviliged port) and it needs to have the vmnetd running but is isn't stated in the error:
`
Error response from daemon: Ports are not available: exposing port UDP 127.0.0.1:53 -> 0.0.0.0:0: failed to connect to /var/run/com.docker.vmnetd.sock: is vmnetd running?: dial unix /var/run/com.docker.vmnetd.sock: connect: no such file or directory
`

**in the docs stated the following:** 
If the privileged helper process is not running, Docker Desktop prompts the user for authorization to run it under [launchd](https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPSystemStartup/Chapters/CreatingLaunchdJobs.html).

The priviliged port helper vmnet deamon needs to be installed when a prompt pops up and give permission.

### Actual behavior

The prompt for installing the deamon never pops up and throws this error: 
`
Error response from daemon: Ports are not available: exposing port UDP 127.0.0.1:53 -> 0.0.0.0:0: failed to connect to /var/run/com.docker.vmnetd.sock: is vmnetd running?: dial unix /var/run/com.docker.vmnetd.sock: connect: no such file or directory
`

This happens every restart of docker for mac

### Information

  - macOS Version: Ventura 13.1
  - Intel chip or Apple chip: M1 pro
  - Docker Desktop Version: v4.16.0

### Output of `/Applications/Docker.app/Contents/MacOS/com.docker.diagnose check`
```
/Applications/Docker.app/Contents/MacOS/com.docker.diagnose check
Starting diagnostics

[PASS] DD0027: is there available disk space on the host?
[PASS] DD0028: is there available VM disk space?
[PASS] DD0018: does the host support virtualization?
[PASS] DD0001: is the application running?
[PASS] DD0017: can a VM be started?
[PASS] DD0016: is the LinuxKit VM running?
[PASS] DD0011: are the LinuxKit services running?
[PASS] DD0004: is the Docker engine running?
[PASS] DD0015: are the binary symlinks installed?
[PASS] DD0031: does the Docker API work?
[PASS] DD0013: is the $PATH ok?
[PASS] DD0003: is the Docker CLI working?
[PASS] DD0038: is the connection to Docker working?
[PASS] DD0014: are the backend processes running?
[PASS] DD0007: is the backend responding?
[PASS] DD0008: is the native API responding?
[PASS] DD0009: is the vpnkit API responding?
[PASS] DD0010: is the Docker API proxy responding?
[SKIP] DD0030: is the image access management authorized?
[PASS] DD0033: does the host have Internet access?
[PASS] DD0018: does the host support virtualization?
[PASS] DD0001: is the application running?
[PASS] DD0017: can a VM be started?
[PASS] DD0016: is the LinuxKit VM running?
[PASS] DD0011: are the LinuxKit services running?
[PASS] DD0004: is the Docker engine running?
[PASS] DD0015: are the binary symlinks installed?
[PASS] DD0031: does the Docker API work?
[PASS] DD0032: do Docker networks overlap with host IPs?
No fatal errors detected.
```

### Steps to reproduce the behavior

  1. install docker desktop for mac 4.16.0
  2. stop and close your docker, then start it again
  3. docker-compose.yml > 
  ```yml
version: "3.9"
services:
  dnsmasq:
    container_name: dnsmasq
    image: rollupdev/dnsmasq
    ports:
    - mode: ingress
      host_ip: 127.0.0.1
      target: 53
      published: "53"
      protocol: udp
    restart: always
```
  4. docker compose up
  5. you got error `Error response from daemon: Ports are not available: exposing port UDP 127.0.0.1:53 -> 0.0.0.0:0: failed to connect to /var/run/com.docker.vmnetd.sock: is vmnetd running?: dial unix /var/run/com.docker.vmnetd.sock: connect: no such file or directory`

**Comment by rbouma** [comment]
**We fixed this temporary manually for now via:**

check if the vmnetd launch service exists in `/Library/PrivilegedHelperTools/com.docker.vmnetd` otherwise copy them over with:
```bash
sudo cp /Applications/Docker.app/Contents/Library/LaunchServices/com.docker.vmnetd /Library/PrivilegedHelperTools/com.docker.vmnetd
```

check if the launch deamon exists in `/Library/LaunchDaemons/com.docker.vmnetd.plist` otherwise create it with:
```bash
sudo tee -a /Library/LaunchDaemons/com.docker.vmnetd.plist > /dev/null <<EOT
        <?xml version="1.0" encoding="UTF-8"?>
        <!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
        <plist version="1.0">
        <dict>
                <key>Label</key>
                <string>com.docker.vmnetd</string>
                <key>Program</key>
                <string>/Library/PrivilegedHelperTools/com.docker.vmnetd</string>
                <key>ProgramArguments</key>
                <array>
                        <string>/Library/PrivilegedHelperTools/com.docker.vmnetd</string>
                </array>
                <key>RunAtLoad</key>
                <true/>
                <key>Sockets</key>
                <dict>
                        <key>Listener</key>
                        <dict>
                                <key>SockPathMode</key>
                                <integer>438</integer>
                                <key>SockPathName</key>
                                <string>/var/run/com.docker.vmnetd.sock</string>
                        </dict>
                </dict>
                <key>Version</key>
                <string>64</string>
        </dict>
        </plist>
EOT
```

Last step is stop the vmnetd service if it is already running, unload, load the service and then start the vmnetd service again
```bash
sudo launchctl stop com.docker.vmnetd
sudo launchctl unload /Library/LaunchDaemons/com.docker.vmnetd.plist
sudo launchctl load /Library/LaunchDaemons/com.docker.vmnetd.plist
sudo launchctl start com.docker.vmnetd
```

this fixed our issue. but it will be better if docker is prompting for installing the deamon to use privileged ports such as 53

**Comment by aiordache** [comment]
Hi @rbouma. 
Looking at the logs you uploaded, we found a failed vmnetd install but no details on the actual error apart from an exit status.
You can use the commands below to remove/install vmnetd:
```
/Applications/Docker.app/Contents/MacOS/install remove-vmnetd
/Applications/Docker.app/Contents/MacOS/install vmnetd
```
Do you mind trying them and check if the install is returning an error and if you can get more details on it please? Thanks!

**Comment by rbouma** [comment]
@aiordache 
```
❯ /Applications/Docker.app/Contents/MacOS/install remove-vmnetd
❯ /Applications/Docker.app/Contents/MacOS/install vmnetd
Error: parsing Info.plist: reading plist /Applications/Docker.app/Contents/Info.Plist: open /Applications/Docker.app/Contents/Info.Plist: no such file or directory
```

Someone at our office has the exact same problem with docker 4.15.0 and 4.16.0 he is running an intel mac latest Mac os ventura

Normally we downgrade to 4.13.1 and then it works fine but after we install 4.15.0 or 4.16.0 directly or upgrade from the older version it is broken.

**Comment by aiordache** [comment]
Sorry for this bug @rbouma. We'll get a fix in for the next release. Can you maybe duplicate the file and have it with both `.Plist` and `.plist` until then?

**Comment by rbouma** [comment]
@aiordache Yes i have both files in the directory and so far it is running good.

**Comment by mat007** [comment]
We have just released Docker Desktop 4.16.2 with a fix!

**Comment by rbouma** [comment]
@mat007 great! Thanks

**Comment by Powfulbadextech** [comment]
Uploading VID_20880412_121732_145.mp4…

Full p.o.p ceiling designs and TV stand

**Comment by Powfulbadextech** [comment]
![IMG_20230131_023320_996](https://user-images.githubusercontent.com/126372576/221394407-7d66c4fb-9eb7-4fde-a0d2-9a61d04808b8.jpg)
Full p.o.p ceiling design

**Comment by monaw** [comment]
i know this is closed but in case this helps anyone...in my case (MacOS 13.4, Docker Desktop Version 4.20.1), the steps mentioned above [by rbouma, updated on 1-16](https://github.com/docker/for-mac/issues/6677#issuecomment-1382313815) was close but didn't work.  these steps worked for me:

1. /Applications/Docker.app/Contents/MacOS/install remove-vmnetd
2. sudo /Applications/Docker.app/Contents/MacOS/install vmnetd

**Comment by jonmann20** [comment]
same as @monaw was happening for me on MacOS 13.4 and Docker Desktop 4.20.1

**Comment by alanzheng88** [comment]
thanks @monaw! that resolved it for me on MacOS 13.4 and docker desktop 4.20.1

**Comment by kay-o** [comment]
Seems several projects have begun recommending [CoLima](https://github.com/abiosoft/colima) when using Docker on Mac. It seems to have solved this issue; trying it out further.

**Comment by annstringer** [comment]
I can confirm that the solution by @monaw worked for me on MacOS 13.5.1 and Docker Desktop 4.23.

**Comment by setswei** [comment]
+1 can also confirm those steps work on 13.4.1 and docker 4.24.2

**Comment by marcosbrasil** [comment]
+1
I can confirm that the solution by @monaw worked for me on MacOS Sonoma 14.0 and Docker Desktop 4.24.2.

**Comment by mogoman** [comment]
worked for me in  Venture upgrading to 4.25.0

**Comment by raoulwegat** [comment]
Thanks @monaw! Also working for me on Docker Desktop 4.25.2 (129061) and Sonoma 14.1.1

**Comment by dhwani07** [comment]
Thanks @monaw  works for me as well macOS: 13.5.2 Ventura, Docker v 4.25

**Comment by slackerzz** [comment]
> i know this is closed but in case this helps anyone...in my case (MacOS 13.4, Docker Desktop Version 4.20.1), the steps mentioned above [by rbouma, updated on 1-16](https://github.com/docker/for-mac/issues/6677#issuecomment-1382313815) was close but didn't work. these steps worked for me:
> 
>     1. /Applications/Docker.app/Contents/MacOS/install remove-vmnetd
> 
>     2. sudo /Applications/Docker.app/Contents/MacOS/install vmnetd

`Docker Desktop 4.26.0 on MacOS Sonoma 14.2`

This helped, but after that, I constantly needed to run `docker system prune`, otherwise, each time I start my containers (via docker compose) I get something like this:
```
Error response from daemon: network 0903dc942d8d6f31628875418f6e873f3872b61fd69c45e108eecd43d162e8ba not found
```

**Comment by mitrol-design** [comment]
One more  for @monaw ’s solution - macOS Sonoma 14.2.1 & Docker Desktop 4.26.1 — Thanks

**Comment by RootProgger** [comment]
Same Problem on MacOS X Ventura, Docker Desktop 4.26.1.
@monaw solution works, but I can't even go to the Sysadmin when i restart Docker-Desktop to call Sudo

**Comment by StephenHinck** [comment]
MacOS Sonoma 14.2.1 with Docker Desktop v4.26.1, same issue and same fix from @monaw worked great. Would be great to get this fixed directly and not require a workaround.

**Comment by taoeffect** [comment]
I am running v4.29.0 on Sonoma 14.4.1, and still had to run @monaw's [commands](#issuecomment-1593787335) to get this to work.

Sounds like this hasn't been fixed? Can the issue be reopened?

**Comment by StephenHinck** [comment]
Updated to v4.31.0 today and it re-broke me. https://github.com/docker/for-mac/issues/6677#issuecomment-1593787335 is still necessary to resolve this issue :-/

**Comment by RootProgger** [comment]
> Updated to v4.31.0 today and it re-broke me. [#6677 (comment)](https://github.com/docker/for-mac/issues/6677#issuecomment-1593787335) is still necessary to resolve this issue :-/

i resolved this issue on my Mac with complete deinstall docker desktop (including remove config files) and reinstall new version.

**Comment by ThePlastic** [comment]
> > Updated to v4.31.0 today and it re-broke me. [#6677 (comment)](https://github.com/docker/for-mac/issues/6677#issuecomment-1593787335) is still necessary to resolve this issue :-/
> 
> i resolved this issue on my Mac with complete deinstall docker desktop (including remove config files) and reinstall new version.

tried, issue still present

**Comment by marcosbrasil** [comment]
I can confirm the same here as @StephenHinck

**Comment by StephenHinck** [comment]
@ThePlastic - reinstall has worked for me so far (and the fix has persisted through one upgrade). Follow the full instructions here: https://docs.docker.com/desktop/uninstall/, including the three terminal commands.

**Comment by KurtRogers** [comment]
Insane. @monaw you are a legend. The commands are _**still**_ needed.

Docker v4.31.0 (153195)
<img width="420" alt="image" src="https://github.com/docker/for-mac/assets/118725785/2d1471a6-f0f0-49e8-ab22-727a5a9d0699">
MacOS 14.5 (23F79)

**Comment by ThePlastic** [comment]
> @ThePlastic - reinstall has worked for me so far (and the fix has persisted through one upgrade). Follow the full instructions here: https://docs.docker.com/desktop/uninstall/, including the three terminal commands.

I didi it, step by step, still need the 2 command listed above....

**Comment by brittonwalker** [comment]
The solve that worked for me was going to Settings -> Advanced and selecting "Allow privileged port mapping".
If it is already selected, which in my case was, I had to uncheck, apply changes, and check again. Solved my issue.

**Comment by carbone** [comment]
> > i know this is closed but in case this helps anyone...in my case (MacOS 13.4, Docker Desktop Version 4.20.1), the steps mentioned above [by rbouma, updated on 1-16](https://github.com/docker/for-mac/issues/6677#issuecomment-1382313815) was close but didn't work. these steps worked for me:
> > 
> > 1. /Applications/Docker.app/Contents/MacOS/install remove-vmnetd
> > 2. sudo /Applications/Docker.app/Contents/MacOS/install vmnetd
> 
> Worked for me as well.

Same. Thanks!

**Comment by kevinquillen** [comment]
This still happens (reported to me several times this week) and yet the issue is locked, and the fix is still to run the two commands.

**Comment by working-name** [comment]
@mat007 maybe a regression but it's not fixed in 4.20.1

**Comment by RalucaElisabetaR** [comment]
> **We fixed this temporary manually for now via:**
> 
> we need to do this after every restart of docker desktop
> 
> check if the vmnetd launch service exists in `/Library/PrivilegedHelperTools/com.docker.vmnetd` otherwise copy them over with:
> 
> sudo cp /Applications/Docker.app/Contents/Library/LaunchServices/com.docker.vmnetd /Library/PrivilegedHelperTools/com.docker.vmnetd
> check if the launch deamon exists in `/Library/LaunchDaemons/com.docker.vmnetd.plist` otherwise create it with:
> 
> sudo tee -a /Library/LaunchDaemons/com.docker.vmnetd.plist > /dev/null <<EOT
>         <?xml version="1.0" encoding="UTF-8"?>
>         <!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
>         <plist version="1.0">
>         <dict>
>                 <key>Label</key>
>                 <string>com.docker.vmnetd</string>
>                 <key>Program</key>
>                 <string>/Library/PrivilegedHelperTools/com.docker.vmnetd</string>
>                 <key>ProgramArguments</key>
>                 <array>
>                         <string>/Library/PrivilegedHelperTools/com.docker.vmnetd</string>
>                 </array>
>                 <key>RunAtLoad</key>
>                 <true/>
>                 <key>Sockets</key>
>                 <dict>
>                         <key>Listener</key>
>                         <dict>
>                                 <key>SockPathMode</key>
>                                 <integer>438</integer>
>                                 <key>SockPathName</key>
>                                 <string>/var/run/com.docker.vmnetd.sock</string>
>                         </dict>
>                 </dict>
>                 <key>Version</key>
>                 <string>64</string>
>         </dict>
>         </plist>
> EOT
> Last step is stop the vmnetd service if it is already running, unload, load the service and then start the vmnetd service again
> 
> sudo launchctl stop com.docker.vmnetd
> sudo launchctl unload /Library/LaunchDaemons/com.docker.vmnetd.plist
> sudo launchctl load /Library/LaunchDaemons/com.docker.vmnetd.plist
> sudo launchctl start com.docker.vmnetd
> this fixed our issue. but it will be better if docker is prompting for installing the deamon to use privileged ports such as 53

has same issue after upgrade and this fixed it ✔️✔️

**Comment by francoiscabrol** [comment]
The process to solve it is now described as a known issue in the docker documentation: https://docs.docker.com/desktop/release-notes/#known-issues

It worked for me.

**Comment by skoskie** [comment]
4.37.2 Checking in. Still an issue. Still resolved by...

> `/Applications/Docker.app/Contents/MacOS/install remove-vmnetd`
> `sudo /Applications/Docker.app/Contents/MacOS/install vmnetd`

**Comment by working-name** [comment]
Yup, still an issue. I gave up hope, so I placed the commands in a function in `$HOME/.zshrc` to make it easier to run every time:

```zsh
function fixdocker() {
  sudo /Applications/Docker.app/Contents/MacOS/install remove-vmnetd
  sudo /Applications/Docker.app/Contents/MacOS/install vmnetd
}
```

All you do is type `fixdocker` like you do `docker compose ...`. Function name = command, so it can be anything you want.

**Comment by rbouma** [comment]
After all the issues I had in the past with Docker Desktop, I finally switched to a new Docker replacement called OrbStack. Check it out if you’re looking for something with fewer problems and much faster performance, including shorter startup times: https://orbstack.dev/

**Comment by skoskie** [comment]
I've heard of Orb and been wanting to try it out, but I need to research how disruptive the switch would be and I've just been way too busy to even think about it lately.  I'm pretty dependent on [DDev](https://ddev.com) and not sure if they are compatible. Definitely glad to hear another positive comment about it though. Docker is so unreliable.

**Comment by rbouma** [comment]
@skoskie yes for me also, used mix of DDEV, Warden, RollDev and its al good and working correctly. OrbStack is also giving you a migration option to seamlessly transfer all projects, containers, volumes and images so you will not loose any data :)

**Comment by bobyui21** [comment]
I am on docker 4.41.2 (191736)

and its happening for me too...

**Comment by working-name** [comment]
> [@skoskie](https://github.com/skoskie) yes for me also, used mix of DDEV, Warden, RollDev and its al good and working correctly. OrbStack is also giving you a migration option to seamlessly transfer all projects, containers, volumes and images so you will not loose any data :)

Biggest issue with OrbStack is it wants to call home to mommy to check on activation on a free license every few minutes. So if I'm ever offline, it forcefully shuts down all my containers and displays a nasty message about not being able to check the license. So, for now, good if you're always connected, and you're comfortable with software making connections behind your back even when you select no updates. https://github.com/orbstack/orbstack/issues/1831

**Comment by lpetri-mmda** [comment]
This worked for me 

> i know this is closed but in case this helps anyone...in my case (MacOS 13.4, Docker Desktop Version 4.20.1), the steps mentioned above [by rbouma, updated on 1-16](https://github.com/docker/for-mac/issues/6677#issuecomment-1382313815) was close but didn't work. these steps worked for me:
> 
> 1. /Applications/Docker.app/Contents/MacOS/install remove-vmnetd
> 2. sudo /Applications/Docker.app/Contents/MacOS/install vmnetd

**Comment by ovi-fabric** [comment]
This happens to me during every single Docker Desktop update. It's been going on for almost two years now. Dissapointing that the issue is closed. Thanks @lpetri-mmda for the fix.

**Comment by bsousaa** [comment]
@ovi-fabric this is a pretty old issue. Can you please submit a new issue?

**Comment by working-name** [comment]
@bsousaa @mat007 The issue is not "pretty old", it's just unaddressed and incorrectly closed.

**Comment by ovi-fabric** [comment]
@bsousaa I get the same error as @rbouma so I don't see the point in opening a new issue. This has not been fixed.

**Comment by maxime-dupuis** [comment]
I had a similar issue:

```
Error response from daemon: ports are not available: exposing port TCP 127.0.0.1:443 -> 127.0.0.1:0: failed to connect to /var/run/com.docker.vmnetd.sock: is vmnetd running?: dial unix /var/run/com.docker.vmnetd.sock: connect: no such file or directory
```

This resolved it for me:
 
- Docker Desktop
- Settings (gear icon)
- Advanced
- Check this setting:
  
    ```
    Allow privileged port mapping (requires password)
      Starts the privileged helper process which binds privileged ports that are between 1 and 1024.
  ```

**Comment by martin-podlubny** [comment]
Hey folks 👋 I've been digging into this issue and wanted to share what I think is going on:

**TL;DR:** To bind privileged ports you need to have the `vmnetd` LaunchDaemon running. Like @maxime-dupuis suggested a while back, the fix is to enable the "Allow privileged port mapping" setting in Docker Desktop → Settings → Advanced, which makes Docker Desktop manage the `vmnetd` LaunchDaemon. 

Manually running `install vmnetd` is possible, but not something officially supported, which is why it keeps breaking after some DD updates.

### Longer explanation:

Binding privileged ports (1-1024) to 127.0.0.1 requires root privileges on macOS. Docker Desktop handles this through `vmnetd`, which runs as a macOS LaunchDaemon. The "official" way to manage this is via the "Allow privileged port mapping" setting.

I was able to reproduce the error, where without the setting enabled, if I try to bind a privileged port, I get the error. I can make it work by manually running `/Applications/Docker.app/Contents/MacOS/install vmnetd`, but that bypasses Docker Desktop's management of this LaunchDaemon. And then some DD updates break it and remove the manually-installed LaunchDaemon. Not all updates, just some:

* DD updates that don't include an update to `vmnetd` will leave this manual setup intact and everything keeps working.
* DD updates that include an update to `vmnetd` remove my manually-installed LaunchDaemon if the setting is unchecked (because DD treats the setting as the source of truth). So the update breaks my manual workaround.

And then yeah, re-running the `install vmnetd` after those updates that remove the manually-installed LaunchDaemon makes it work again. But overall, better to enable the "Allow privileged port mapping" setting and let Docker Desktop manage the `vmnetd` LaunchDaemon properly. Then binding privileged ports should work.

As to why the docs hinted at Docker Desktop prompting the user to install the helper first time it's needed (at least that's how I read the original issue description), that I don't know.

Hope this helps clarify things. I think the root of the problem is/was that manually managing `vmnetd` puts you outside the expected path, and DD isn't handling that gracefully. That being said, if you're still hitting issues after enabling the setting, please let me know 👍

**Comment by OpenPj** [comment]
> i know this is closed but in case this helps anyone...in my case (MacOS 13.4, Docker Desktop Version 4.20.1), the steps mentioned above [by rbouma, updated on 1-16](https://github.com/docker/for-mac/issues/6677#issuecomment-1382313815) was close but didn't work. these steps worked for me:
> 
> 1. /Applications/Docker.app/Contents/MacOS/install remove-vmnetd
> 2. sudo /Applications/Docker.app/Contents/MacOS/install vmnetd

I had the same issue with Mac OS Tahoe 26.3 (25D125) and Docker Desktop 4.63.0 and resolved in this way!

**Comment by gm5621702-cpu** [comment]
❯ /Applications/Docker.app/Contents/MacOS/install remove-vmnetd
❯ /Applications/Docker.app/Contents/MacOS/install vmnetd
Error: parsing Info.plist: reading plist /Applications/Docker.app/Contents/Info.Plist: open /Applications/Docker.app/Contents/Info.Plist: no such file or directory
